test/storage_accessible.js

```diff
@@ -1,51 +1,83 @@
+      const length = await instance.getStorageAt(slot, 1)
+      assert.equal(fromHex(length), 2)
 
-    const data = await instance.getStorageAt(keccak([slot]), length)
-    assert.equal(1, fromHex(data.slice(2, 66)))
-    assert.equal(2, fromHex(data.slice(66, 130)))
-  })
-  it('can read mappings', async () => {
-    const instance = await StorageAccessibleWrapper.new()
-    await instance.setQuxKeyValue(42, 69)
-    assert.equal(69, await instance.getStorageAt(keccak([42, await instance.SLOT_QUX()]), 1))
+      const data = await instance.getStorageAt(keccak([slot]), length)
+      assert.equal(1, fromHex(data.slice(2, 66)))
+      assert.equal(2, fromHex(data.slice(66, 130)))
+    })
+    it('can read mappings', async () => {
+      const instance = await StorageAccessibleWrapper.new()
+      await instance.setQuxKeyValue(42, 69)
+      assert.equal(69, await instance.getStorageAt(keccak([42, await instance.SLOT_QUX()]), 1))
+    })
+
+    it('can read structs', async () => {
+      const instance = await StorageAccessibleWrapper.new()
+      await instance.setFoobar(19, 21)
+
+      const packed = await instance.getStorageAt(await instance.SLOT_FOOBAR(), 10)
+      assert.equal(19, fromHex(packed.slice(2, 66)))
+      assert.equal(21, fromHex(packed.slice(66, 130)))
+    })
   })
 
-  it('can read structs', async () => {
-    const instance = await StorageAccessibleWrapper.new()
-    await instance.setFoobar(19, 21)
+  describe.only('executeStaticDelegatecall', async () => {
+    it('can invoke a function in the context of a previously deployed contract', async () => {
+      const instance = await StorageAccessibleWrapper.new()
+      await instance.setFoo(42)
+
+      // Deploy and use reader contract to access foo
+      const reader = await ExternalStorageReader.new()
+      const getFooCall = reader.contract.methods.getFoo().encodeABI()
+      const result = await instance.executeStaticDelegatecall.call(reader.address, getFooCall)
+      assert.equal(42, fromHex(result))
+    })
+
+    it('can simulate a function with side effects (without executing)', async () => {
+      const instance = await StorageAccessibleWrapper.new()
+      await instance.setFoo(42)
+
+      // Deploy and use reader contract to get and replace foo
+      const reader = await ExternalStorageReader.new()
+      const replaceFooCall = reader.contract.methods.replaceFoo(69).encodeABI()
+      let result = await instance.executeStaticDelegatecall.call(reader.address, replaceFooCall)
+      assert.equal(42, fromHex(result))
 
-    const packed = await instance.getStorageAt(await instance.SLOT_FOOBAR(), 10)
-    assert.equal(19, fromHex(packed.slice(2, 66)))
-    assert.equal(21, fromHex(packed.slice(66, 130)))
+      // Make sure foo is not actually changed
+      const getFooCall = reader.contract.methods.getFoo().encodeABI()
+      result = await instance.executeStaticDelegatecall.call(reader.address, getFooCall)
+      assert.equal(42, fromHex(result))
```

Comment:
OK, so the value isn't changed... I guess this means you can temporarily (i.e. within a single transaction) emulate what would or could happen in the event of a state change without every actually changing it... Is this what one might want it for?

Reply:
Here is an example of we would like to do with it in the context of the safe contracts ;) https://github.com/gnosis/safe-contracts/blob/development/contracts/GnosisSafe.sol#L276

Reply:
Cool example. If every smart contract developer picked up this technique, doesn’t this mean half the transactions submitted would be expected to fail?